Improve io-queue cost model

StephanDollberg opened the pull request:
This PR proposes and initiates a discussion to change the io-queue cost
model by making its costing less pessimistic and hence work around cases 
where the io-queue is massively limiting disk performance when io-properties
are applied.

Currently the IO request cost in the io-queue is defined as such:

```math
cost = 1 * X + S * Y
```


X and Y representing the cost of a single IOP and a byte of throughput
are defined as such:

```math
X = 1 / IOPS
```
```math
Y = 1 / TP
```

With `IOPS` and `TP` being the IOPS and throughput measurement values
from io-properties.

This results in a fairly simple model that doesn't really represent how
modern SSDs work. Often the IOP and bandwidth cost isn't additive but
rather independent. As a consequence, they get heavily throttled by the
io-queue (as seen for example in https://github.com/scylladb/seastar/issues/2771). In such cases we might only reach 
around 60% of the IOPS that are specified in io-properties.

To work around this issue we introduce a more complete cost model which
fully models the above equation based on the two measurements that we
have in io-properties.

The only two additional values that we need are the IOP sizes in use
during both the IOP test (`IOPS_IOP_SIZE`) and throughput test
(`TP_IOP_SIZE`) from iotune. Possibly these can be provided by iotune
as well (implemented in the first commit). For correctness and performance 
of the model it's important that these two values limit our min IOP size
(`IOPS_IOP_SIZE` as as seen in the IOPS test) and our max IOP size 
(`TP_IOP_SIZE` as seen in the throughput test).

With those two additional values we can solve the system of two equations
and unknowns based on our two iotune measurements:

```math
IOPS * X + IOPS * \text{IOPS\_IOP\_SIZE} * Y = 1
```
```math
\frac{TP}{\text{TP\_IOP\_SIZE}} * X + TP * Y = 1
```

Again X represents the cost of an IOP and Y the weight of a byte of
throughput. However, in contrast to the simple model a factor is applied
to either side such that each doesn't contribute in full. We can
solve the above to:


```math
X = \frac{\frac{1}{\text{IOPS}} - \frac{\text{IOPS\_IOP\_SIZE}}{\text{TP}}}{1 - \frac{\text{IOPS\_TP} \cdot \text{IOPS\_IOP\_SIZE}}{\text{TP}}}
```
```math
Y = \frac{1 - \text{X} \cdot \text{IOPS\_TP}}{\text{TP}}
```

leaving us with more accurate cost approximations than the simple one
mentioned above that always weighs an IOP/byte at full cost.

This partially solves the issue described above of the io-queue severly
limiting performance as it's a lot less pessimistic. 

Below is some data from some experiments using io-tester with the
following config (with reqsize 4kB, 16kB and 128kB) respectively:

```
- name: highprio
  shards: [0]
  type: randwrite
  shard_info:
    parallelism: 32
    reqsize: 4kB|16kB|128kB
    shares: 1000
    think_time: 0

- name: lowprio
  shards: [0]
  type: randwrite
  shard_info:
    parallelism: 32
    reqsize: 4kB|16kB|128kB
    shares: 100
    think_time: 0
```

It's important that we use two separate priority groups here to show
that scheduling is still working as expected (without io-properties we
generally get the best performance but scheduling doesn't work at all).

We test on four different systems (all on AWS):

 - m7gd.4xlarge: 135k IOPS (at 4k IOP size) / 604MB/s (at 128k IOP size)
 - standardish EBS: 3k IOPS / 125 MB/s
 - high throughput EBS: 4k IOPS / 1000MB/s
 - m8gd.4xlarge: 135k IOPS (at 4k IOP size) / 907MB/s (at 128k IOP size)

We use `min_iop_size=4k` for all and `max_iop_size=128k` for the dedicated
instances and 256k for EBS.

Below is the data across all systems. The data is in MB/s and split
into the separate priority groups (again to show that scheduling still
works) with the total of both next to it.

NOIP: Without using io-properties
IPSM: Using io-properties with the current simple model
IPFM: Using io-properties with the full model

m7gd.4xlarge:

|       | 4k        |           |         | 16k       |           |         | 128k      |           |         |
|-------|-----------|-----------|---------|-----------|-----------|---------|-----------|-----------|---------|
|       | High Prio | Low Prio  | Total   | High Prio | Low Prio  | Total   | High Prio | Low Prio  | Total   |
| NOIP  | 303       | 287       | 591     | 299       | 292       | 591     | 295       | 295       | 590     |
| IPSM  | 262       | 28        | 290     | 432       | 44        | 476     | 528       | 53        | 581     |
| IPFM  | 477       | 56        | **533**     | 540       | 55        | **595**     | 544       | 55        | 599     |

On m7gd at small write sizes is where we see a massive difference.
Because m7gd basically behaves as if everything can be costed in terms
of 4k IOPS the full model behaves a lot better and we get basically 80%
more performance.

m8gd.4xlarge:

|       | 4k        |           |         | 16k       |           |         | 128k      |           |         |
|-------|-----------|-----------|---------|-----------|-----------|---------|-----------|-----------|---------|
|       | High Prio | Low Prio  | Total   | High Prio | Low Prio  | Total   | High Prio | Low Prio  | Total   |
| NOIP  | 319       | 290       | 609     | 445       | 441       | 886     | 443       | 443       | 886     |
| IPSM  | 331       | 33        | 364     | 599       | 60        | 660     | 786       | 79        | 865     |
| IPFM  | 486       | 105       | **591**     | 741       | 75        | **816**     | 823       | 83        | 905     |

Similarly on m8gd we see quite the performance improvement. m8gd has
this bimodal distribution where at 8k IOP size you get ~80% more
throughput than at 4k IOP. Hence we do also see an improvement at 16k
IOP size though it doesn't reach the full throughput.

standard EBS:

|       | 4k        |           |         | 16k       |           |         | 128k      |           |         |
|-------|-----------|-----------|---------|-----------|-----------|---------|-----------|-----------|---------|
|       | High Prio | Low Prio  | Total   | High Prio | Low Prio  | Total   | High Prio | Low Prio  | Total   |
| NOIP  | 7         | 7         | 13      | 27        | 26        | 53      | 64        | 63        | 127     |
| IPSM  | 11        | 1         | 12      | 34        | 3         | 37      | 72        | 7         | 79      |
| IPFM  | 12        | 1         | 13      | 39        | 4         | 43      | 91        | 9         | **100**     |

On a fairly standard EBS drive we do also see some improvement even at
128k IOP size. These "in-between" disk types show where the model is
still performing the worst. Even with the full model we only get around
80% of the throughput that the disk provides.

high throughput EBS:

|       | 4k        |           |         | 16k       |           |         | 128k      |           |         |
|-------|-----------|-----------|---------|-----------|-----------|---------|-----------|-----------|---------|
|       | High Prio | Low Prio  | Total   | High Prio | Low Prio  | Total   | High Prio | Low Prio  | Total   |
| NOIP  | 8         | 8         | 16      | 32        | 32        | 64      | 259       | 253       | 512     |
| IPSM  | 14        | 1         | 16      | 55        | 5         | 60      | 311       | 31        | 343     |
| IPFM  | 15        | 1         | 16      | 58        | 6         | 64      | 465       | 47        | **512**     |

On throughput sided EBS setups we again get a fairly large improvement
of a 50% throughput (this time at large IOP sizes).

There is an open question as to how `min_iop_size` and `max_iop_size` should be
provided. This PR provides a commit where they are being set (fairly
pessimistically) directly and by default by iotune. As an alternative this could
happen only conditionally based on a cli flag. Which would leave the full model
off by default. Alternatively they could be fully provided by the user. Chosing
them correctly is important for the accuracy of the full model.

Also we could split the two parameters also across read and write which I haven't 
done in this first draft.

avikivity commented:
> Often the IOP and bandwidth cost isn't additive but rather independent. 

Is this a problem? bandwidth is issued using very large requests, so the IOPS-cost is small.

4/128k = 1/32, so we add steal 3% ftom the IOPS budget.

avikivity commented:
> We test on four different systems (all on AWS):
> 
> * m7gd.4xlarge: 135k IOPS (at 4k IOP size) / 604MB/s (at 128k IOP size)


604MB/s @ 128k = 4832 IOPS. So even if the large-write process is running full-tilt, it's only stealing 4.8k IOPS from the 4k process.

I think I see. The bandwidth itself is stealing IOPS, even though this particular disk does not trade bandwidth for IOPS. The original model equations came from the triangles discovered by diskplorer (https://github.com/scylladb/diskplorer), but it's clearly not a fit-all model.

avikivity commented:
Please also share latency results. For those, the think_time of the high-share process should be higher (so it lets the low-share process "steal" disk bandwidth) but the I/O scheduler should still guarantee it low latency.

Please also run diskplorer on those disks (and share results via a pull request) so we can see how it works.

m7gd.4xlarge disk is likely shared (it's probably a 1900GB disk shared across two instances, or maybe even larger) so you could be seeing a quiet neighbor effect. If the other tenant isn't hammering the disk, you're using more than your fair share of the controller bandwidth, and when scaling to a large instance size you'd see worse performance.

xemul commented:
> $$ I O P S ∗ X + I O P S ∗ IOPS\\_IOP\\_SIZE ∗ Y = 1 $$
> $$ \frac {TP} {TP\\_IOP\\_SIZE} ∗ X + T P ∗ Y = 1 $$

Please, explain where the equations come from. Given

$$ X = 1 / I O P S $$
$$ Y = 1 / T P $$

They relax into

$$ 1 + \frac {IOPS} {TP} ∗ IOPS\\_IOP\\_SIZE = 1 $$
$$ \frac {TP} {IOPS∗TP\\_IOP\\_SIZE} + 1 = 1$$

which makes very little sense

StephanDollberg commented:
> The bandwidth itself is stealing IOPS, even though this particular disk does not trade bandwidth for IOPS.

Yeah exactly this is the problem. On m7gd the biggest impact is on the IOPS side where the bandwidth cost steals/has a large impact on the IOPS cost and hence we only reach a fraction of the available IOPS.

The opposite problem can be seen on the "highthroughput EBS" case. There the IOPS steal bandwidth and we can only achieve a fraction of the available bandwidth.

> Please also run diskplorer on those disks (and share results via a pull request) so we can see how it works.

Sure I will look into this. Seems like that needs to run for a while so won't be immediate.

> m7gd.4xlarge disk is likely shared (it's probably a 1900GB disk shared across two instances, or maybe even larger) so you could be seeing a quiet neighbor effect. If the other tenant isn't hammering the disk, you're using more than your fair share of the controller bandwidth, and when scaling to a large instance size you'd see worse performance.

It's likely shared yes. Though from our testing on m7gd and m8gd this doesn't seem to make a difference and they keep behaving the same once you reach the "full-disk" instance type sizes. As a side note these disks are also fully duplex as all the other disks in the newer instance types on AWS. 

> Please also share latency results. For those, the think_time of the high-share process should be higher (so it lets the low-share process "steal" disk bandwidth) but the I/O scheduler should still guarantee it low latency.

Sure I can do that. I tried something like you suggested.

```
- name: highprio
  shards: [0]
  type: randwrite
  shard_info:
    parallelism: 32
    reqsize: 4kB
    shares: 1000
    think_time: 500us

- name: lowprio
  shards: [0]
  type: randwrite
  shard_info:
    parallelism: 32
    reqsize: 4kB
    shares: 100
    think_time: 0
```

  | high prio |   | low prio |  
-- | -- | -- | -- | --
  | TP MB/s | p50 us | TP MB/s | p50 us
NOIP | 139 | 277 | 451 | 254
IPSM | 200 | 52 | 22 | 1684
IPFM | 178 | 75 | 355 | 307

Happy to give others a spin as well.

> Please, explain where the equations come from. Given

Sorry I guess I wasn't clear. So these:

```math
X = 1 / IOPS
```
```math
Y = 1 / TP
```

are the current cost factors that are in use and they are unrelated to the new ones (shown later).

Though you can actually arrive at them if you drop the throughput factor for the IOPS measurement and the IOPS factor for the throughput measurement from the full formula (sorry it's a screenshot as `\cancel` doesn't seem to work on github):

<img width="250" alt="image" src="https://github.com/user-attachments/assets/5038347e-fef0-4fe0-846c-8d6fd34c534a" />

This is why in the simple formula both IOP and throughput cost contribute in full and are not weighted.

Solving for X and Y in the full formula gives the solutions also posted above which are more complicated.

StephanDollberg commented on doc/io-properties-file.md:
```diff
@@ -24,6 +24,12 @@ Additionally the following optional properties can be added:
 
 * `read_saturation_length`: read buffer length to saturate the device throughput
 * `write_saturation_length`: write buffer length to saturate the device throughput
+* `min_iop_size`: The minimum IOP size the application will issue.
```
Yes we should add it there as well. It should be in both places as the values specified here are also important for the IO sizes chosen by the application. I will add it depending on whether we make iotune emit those values or not.

avikivity commented:
I'd like to see p95 and p99 latency in addition to p50.

It doesn't help achieving more throughput if the latency goes to pieces.

(the scheduler should be / is sensitive to io-latency-goal so the application can choose the tradeoff)

avikivity commented:
The entire point of the scheduler is to throttle low-priority high-concurrency demand so that latency remains low. If the input concurrency is low, we don't need a scheduler.

Often the input concurrency is infinite. Consider a compaction job with read-ahead and write-behind. It can generate any amount of concurrency (even in a single thread) as long as it has enough CPU.

The lowprio job should not be capped. The goal is to see the scheduler throttle it.

StephanDollberg commented:
> The entire point of the scheduler is to throttle low-priority high-concurrency demand so that latency remains low. 

Sure

> If the input concurrency is low, we don't need a scheduler.

> Often the input concurrency is infinite. Consider a compaction job with read-ahead and write-behind. It can generate any amount of concurrency (even in a single thread) as long as it has enough CPU.

Are you referring to io concurrency or io demand here? I don't think concurrency can be infinite (the io-queue would grow to infinite size) and similarly I do think we do also need a scheduler at low io concurrency. For io demand yes I agree. 

My point is that with higher io concurrency you can make application observed latencies arbitrarily bad as there is too much concurrency in the disk/io-queue for it to serve it fast enough and hence latency goes up. A faster disk/io-queue will look better here as it gets through the queue faster.

> The lowprio job should not be capped. The goal is to see the scheduler throttle it.

This is effectively the original test then as io demand from the low prio job is infinite.

In any case a similar example, this time using rps instead of think time (which I find hard to reason about):

```
- name: highprio
  shards: [0]
  type: randwrite
  shard_info:
    parallelism: 8
    reqsize: 4kB
    shares: 1000
    think_time: 0
    rps: 2000

- name: lowprio
  shards: [0]
  type: randwrite
  shard_info:
    parallelism: 32
    reqsize: 4kB
    shares: 100
    think_time: 0
```


  | high prio |   |   |   |   | low prio |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
  | TP MB/s | p50us | p95us | p99us | p999us | TP MB/s | p50 us | p95 us | p99 us | p999us
IPSM | 59 | 64 | 156 | 2413 | 3286 | 222 | 540 | 655 | 2906 | 3563
IPFM | 55 | 56 | 428 | 2614 | 3365 | 470 | 228 | 284 | 2552 | 3583


> btw there's no point measuring below a full disk. The result depends on whether you have a noisy or quiet neighbor. The scheduler can't tell, so it must assume the worst.

I don't follow how this is related. The scheduler should behave as per io-properties?

avikivity commented:
> > The entire point of the scheduler is to throttle low-priority high-concurrency demand so that latency remains low.
> 
> Sure
> 
> > If the input concurrency is low, we don't need a scheduler.
> 
> > Often the input concurrency is infinite. Consider a compaction job with read-ahead and write-behind. It can generate any amount of concurrency (even in a single thread) as long as it has enough CPU.
> 
> Are you referring to io concurrency or io demand here? I don't think concurrency can be infinite (the io-queue would grow to infinite size) and similarly I do think we do also need a scheduler at low io concurrency. For io demand yes I agree.

For demand.

> 
> My point is that with higher io concurrency you can make application observed latencies arbitrarily bad as there is too much concurrency in the disk/io-queue for it to serve it fast enough and hence latency goes up. A faster disk/io-queue will look better here as it gets through the queue faster.
> 
> > The lowprio job should not be capped. The goal is to see the scheduler throttle it.
> 
> This is effectively the original test then as io demand from the low prio job is infinite.
> 

All the tests I saw had parallelism of 8 and 32, that's self-limiting for these disks.


> In any case a similar example, this time using rps instead of think time (which I find hard to reason about):


rps is better than think time.

> 
> ```
> - name: highprio
>   shards: [0]
>   type: randwrite
>   shard_info:
>     parallelism: 8
>     reqsize: 4kB
>     shares: 1000
>     think_time: 0
>     rps: 2000
> 
> - name: lowprio
>   shards: [0]
>   type: randwrite
>   shard_info:
>     parallelism: 32
>     reqsize: 4kB
>     shares: 100
>     think_time: 0
> ```
> 
>  	high prio	 	 	 	 	low prio	 	 	 	 
>  	TP MB/s	p50us	p95us	p99us	p999us	TP MB/s	p50 us	p95 us	p99 us	p999us
> IPSM	59	64	156	2413	3286	222	540	655	2906	3563
> IPFM	55	56	428	2614	3365	470	228	284	2552	3583
> > btw there's no point measuring below a full disk. The result depends on whether you have a noisy or quiet neighbor. The scheduler can't tell, so it must assume the worst.
> 
> I don't follow how this is related. The scheduler should behave as per io-properties?

Yes. But the disk will give good results if you have a quiet neighbor, even if the scheduler misbehaves.

It's similar to setting artificially  low io_properties.conf. Yes, you'll get good latency, but it's because the disk is much better than you told the scheduler. The price is that bandwidth is artificially low.

xemul commented:
Presumably [my understanding](https://github.com/scylladb/seastar/pull/2801#issuecomment-2983397822) is correct, at least the change in `internal::request_tokens` supports that:

```patch
- return double(m.weight) / cfg.req_count_rate             + double(m.size) * (dnl.length() >> io_queue::block_size_shift) / cfg.blocks_count_rate;
+ return double(m.weight) / cfg.req_count_rate * f.iop + double(m.size) * (dnl.length() >> io_queue::block_size_shift) / cfg.blocks_count_rate * f.throughput;
```

@StephanDollberg , I think that the run-time math should be left intact, respectively cfg.req_count_rate and cfg.blocks_count_rate should be respectively evaluated in src/core/reactor.cc.

I also have concerns about max_iop_size, but I'll leave it comment in the patch itself.

xemul commented on doc/io-properties-file.md:
```diff
@@ -24,6 +24,12 @@ Additionally the following optional properties can be added:
+* `min_iop_size`: The minimum IOP size the application will issue.
+    Should match (or be smaller) than IOP size used in the IOPS test.
+    Used in the io-queue cost model.
+* `max_iop_size`: The maximum IOP size the application will issue.
```
Will or should? Currently max io size is advertised via  io_queue::get_request_limits() and it doesn't look like the values returned from there are anyhow affected by that io-properties parameter, are they?

xemul replied:
I see. Well yes, I agree that what iotune uses vs what it adverises up the stack should not necessarily match. Provided [my understanding](https://github.com/scylladb/seastar/pull/2801#discussion_r2179938259) is (still) correct, I would expect values and comments that describe that new "line on iops;bandwidth diskplorer plane"

xemul commented on apps/iotune/iotune.cc:
```diff
@@ -104,6 +104,7 @@ struct evaluation_directory {
     unsigned _force_io_depth;
     uint64_t _available_space;
     uint64_t _min_data_transfer_size = 512;
+    uint64_t _max_iop_size = 256 * 4096;
```
Despite patch commit assumes, this value is pure abstract in a sense that -- iotune still uses some _other_ request size for its bandwidth tests. Am I wrong?

StephanDollberg replied:
> Despite patch commit assumes, this value is pure abstract in a sense that -- iotune still uses some other request size for its bandwidth tests. Am I wrong?

Yes you are (partially) right, iotune still submits 1MB requests unconditionally. It's at the block layer where requests will then get split up into multiple requests as per this value.

We should probably change iotune to respect this in any case but I didn't want to overload this PR initially.

xemul commented on src/core/reactor.cc:
```diff
@@ -4228,6 +4316,19 @@ class disk_config_params {
+            p.min_iop_size && p.max_iop_size) {
+
+            std::tie(cfg.write_iop_factor, cfg.write_throughput_factor) = calculate_full_cost_formula_factors(true, p);
+            std::tie(cfg.read_iop_factor, cfg.read_throughput_factor) = calculate_full_cost_formula_factors(false, p);
```
Please, note, there's "rate_factor" thing in config, which now interferes with these new factors and make things very complicated. Like this

First, we have iops_max and bandwidth_max -- the points at which the "diskplorer line" crosses the x-y axis.
Then the rate_factor is applied this shifting (scaling?) the line along the 45-degrees vector.
Now new parameters are applied and shift the individual corner points each by its own factor.

So far I've no good ideas what to do about it, it's just an observation

xemul replied:
After a bit more thinking, I recommend making rate factor and new parameters mutually exclusive. I.e. -- when both are present in the io-properties, the mountpoint_params creator code should throw configuration exception. They both tune that "famous" line on the iops:bw plane, and should be two different ways of doing it, not two incremental updates to the default line.

StephanDollberg commented:
> @StephanDollberg , I think that the run-time math should be left intact, respectively cfg.req_count_rate and cfg.blocks_count_rate should be respectively evaluated in src/core/reactor.cc.

Right so you mean `internal::request_tokens` shouldn't change and instead we should multiply things a priori?

We can do that, it would then basically be this in the new if statement in `generate_config`:

```diff
+            cfg.req_count_rate /= cfg.read_iop_factor;
+            cfg.blocks_count_rate /= cfg.read_throughput_factor;
+            cfg.disk_blocks_write_to_read_multiplier *= cfg.write_throughput_factor / cfg.read_throughput_factor;
+            cfg.disk_req_write_to_read_multiplier *= cfg.write_iop_factor / cfg.read_iop_factor;
```

I had it done with the separate factor because I find it easier to understand initially but happy to do whichever way.

avikivity commented:
> > OK, while initial formula found the line (on diskplorer areas) at {iops_max, 0} and {0, bandwidth_max} points
> 
> This part is a bit confusing to me. Which disk diskplorer charts? The ones in the README.md are all "ready latency vs write bandwidth", which is largely evaluating the "read vs write" tradeoff, not the "iops vs bandwidth" tradeoff, i.e., it is mostly assessing whether the disk is full-duplex or not (and the answer seems to be quite mixed here with modern disks).

Diskplorer found the triangle, which let to the IOPS/IOPSmax + BW/BWmax < 1 equation.


> 
> For evaluating the cost function I would kind of expect to start with pure write or pure read tests, and then compare IOPS vs BW.

The axes on diskplorer charts represent pure random read and almost-pure sequential write (with a sprinkling of random reads to have latency measurement).

> 
> The main thing though is there aren't really measured points like `{iops_max, 0}` are there? At `iops_max` you will have non-zero BW, and you should not neglect it as it can be high in practice. Of course, you can set the BW to 0 on the "x axis" workload, but then you should still add the BW contributed by the "y axis" workload to get the total BW.
> 
> The other thing is that the model is degenerate below a certain size, e.g., even if there is an IOPS/BW tradeoff in the main interesting region of the disk, once you drop below 4K on every SSD I've seen there is no additional increase in IOPS: it either decreases, or on some disks that implement RMW efficiently it might be the same as 4K. So if you aren't doing IOs smaller than 4K, I think it's much better just to measure at 4K as that will give a good model at and endpoints and reasonable in the middle. If you _do_ do a lot of IO smaller than 4K, there is no completely obvious choice (you introduce error one way or another), though if it was important you could fix it by rounding up IOs to 4K in the scheduler model and running iotune at 4K.

It depends on whether the bottleneck is PCIe or the disk media, and whether the disk itself can read 512 bytes or not. We should measure those things rather than assume them. But I agree most disks need to use 4k.


> 
> There is some of kind of similar degeneracy at the high block size end but I think it matters much less because of how the math works out.

We neglected the block size contribution at max IOPS purposefully, as an addition cushion, but maybe it's too generous.

ScyllaDB is much more sensitive to latency than it is to maximum bandwidth. A query consumes tons of CPU cycles, so it rarely chokes on maximum bandwidth, but if a background job starts hurting latency, it is felt.

avikivity commented:
> t 4K granula



> > Trying to remember where this came from, I think it's commitlog where we reduced the filesystem block size to reduce commitlog bandwidth where tiny commits caused bandwidth amplification which in turn ate performance.
> 
> I recall some of those scylladb threads, though I thought the concern was more about disk usage amplification than BW amplification.
> 
> BTW in Redpanda we have a similar case in the primary raft logs where we are often writing small chunks (whose size is up to the client, not us) and where latency and throughput are important: we handle it by repeatedly appending and writing at 4K granularity, but we there is no space or BW amplification as each write goes next the previous one without any padding in between them. So if the appends are 1000 bytes in size we just end up writing the same 4K block 4x each with a bit more logical payload each time. I guess the main downside is that you can only have 1 of these writes in flight at once, so when the writes are arriving quickly the 2nd write may block: but this effect is limited as long as you merge all waiting writes, as then the loss in parallelism is made up by with larger writes.

I'm concerned about overwrites since they can cause write amplification, but perhaps this is misplaced: first, they're likely to be absorbed by the disk cache, and second if you pad you get the same write amplification. /cc @elcallio 

Not having concurrent writes to the commitlog is the least of my worries.


> 
> > We should stick close to the physics.
> 
> Sure. I.e., measurement.
> 
> > For writes, we should use the physical block size reported by the disk (/sys/block/*/queue/physical_block_size).
> 
> When you say "use", for what purpose? Do you mean when choosing the endpoints of the triangle/where `iotune` should perform its measurement?

Yes. And likely for runtime too. Not cause internal amplification by forcing the disk to read-modify-write (although again, that's likely absorbed by the disk caches).

> 
> `physical_block_size` seems to be always 512 on AWS NVMe, and some experiments seem to show that in fact they do behave as if they support "native" 512b IOPS, i.e., when mixed IOPS sizes are in play and the disk hits its throughput limit, 512b writes seem to cost only 1/8th as much as 4K writes (i.e., at the throughput bottleneck point, it is not simply "as if" they are rounded up to 4K.
> 
> What matters most is how the application does IO though, I think? If we only do 4K writes or larger, measuring at 512b just introduces a significant and unnecessary error.


Why would it introduce an error? If the model is correct, then it should predict the cost of a 4k write correctly.

Why would the application only do 4k writes if the disk supports a lower granularity? It's wasteful.

avikivity commented:
> @avikivity wrote:
> 
> > Why would it introduce an error? If the model is correct, then it should predict the cost of a 4k write correctly.
> 
> The model is not correct. At least for many cloud SSDs it is quite far off, in several different aspects:

Then let's correct it, rather than changing random parameters.

> 
> * Assumption that IOPS cost and tput costs sum to a final total cost

Why not? That's exactly the diskplorer triangle. ax + by = const.

Remember diskplorer doesn't find the maximum achievable throughput and IOPS; rather the maximum achievable combination that gives reasonable latency.



> * Assumption that a single smooth/linear formula holds across the range, especially the part below 4K (in fact, there is a step change in behavior below 4K)

Then let's add an additional data point to iotune (and perhaps diskplorer).

The model doesn't just follow from diskplorer results, it can be understood as the disk controller reaching its limits. A change in behavior at 4096 bytes can also be understood as the controller handling part of the request as a 4096 byte read and other parts of the request as a 512 byte read.

> * Let us not even talk yet about full/half-duplex aspects when mixing reads and writes

Well, diskplorer measured mixed reads/writes and saw that they directly affect reach other. So at least for those disks, the bottleneck is in a shared path. If anything mixed read/read would be worse.


> * Wrong calculation of IOPS and tput costs from iotune input
> 
> That's not intended as a criticism of the model, just an observation.

Criticism of the model is welcome, and even more if it comes with corrections.

In the end this model is a guess inspired by the diskplorer triangles.

> 
> Given that, this change only tries to improve the last point, and it is very important then to choose the endpoints of the measured region to give the least error in the region you care about, because the other inaccuracies still exist.

If there is indeed a behavioral change between 512 and 4096 bytes, we should extend the model to be a polygon rather than a straight line and measure the inflection points.


> 
> > Why would the application only do 4k writes if the disk supports a lower granularity? It's wasteful.
> 
> One good reason would be because it is writing to XFS disks with the default block size of 4K, where performance is very bad for smaller writes for "filesystem reasons" rather than disk reasons.

This should not be the case, XFS improved sub-block overwrite performance some years ago.

Note ScyllaDB formats XFS with 1k block size (since the XFS improvement happened at the same time we noticed the problem, and could not wait for kernel propagation).

Perhaps the iomap context switch you found is hurting us, we should look at addressing it. I'm guessing that splitting the iomap wakeup to have an inline check if a wakeup is needed, then do the wakeup conditionally should be possible, since that's what XFS did previously.

You should not be using appending writes, they are too terrible.

> 
> Another reason might be because the writes just happen to be at that size due to the design of the application or typical workloads. There may be writes smaller than 4K but uncommon.

They're common for us.

Why would 4k writes be more common? Reasonable sizes are "as small as possible" to prevent amplification (of course, for some workloads those could be large), or 8k/16k random writes for btree style writes.

> 
> Or you might have a ton of 512b writes: it's application dependent. That's why we suggest leaving it up to the application.

ScyllaDB will have 1024b writes in some workloads, and larger writes in others, it's workload dependent within the same application.

StephanDollberg commented:
> rather than changing random parameters.

I mean the changes aren't entirely random. It's more like making the most out of the current model and measurements we already have.

> If there is indeed a behavioral change between 512 and 4096 bytes, we should extend the model to be a polygon rather than a straight line and measure the inflection points.

The true model model is likely some step function with many steps. Travis recently did some sub 4k testing and there seems to be quite varying behaviour. See for example on m8gd.large and m8gd.8xlarge (Travis has some more data but currently OOOish): 

.large (partial disk):
```
BS: 512 IOPS: 16770 BW: 8 MB/s
BS: 1024 IOPS: 16770 BW: 16 MB/s
BS: 1536 IOPS: 16770 BW: 25 MB/s
BS: 2048 IOPS: 16770 BW: 33 MB/s
BS: 2560 IOPS: 16770 BW: 41 MB/s
BS: 3072 IOPS: 16770 BW: 49 MB/s
BS: 3584 IOPS: 12613 BW: 43 MB/s
BS: 4096 IOPS: 16770 BW: 66 MB/s
BS: 4608 IOPS: 16770 BW: 74 MB/s
BS: 5120 IOPS: 16770 BW: 82 MB/s
BS: 5632 IOPS: 16770 BW: 90 MB/s
BS: 6144 IOPS: 16770 BW: 98 MB/s
BS: 6656 IOPS: 16770 BW: 106 MB/s
BS: 7168 IOPS: 16335 BW: 112 MB/s
BS: 7680 IOPS: 14684 BW: 108 MB/s
BS: 8192 IOPS: 13765 BW: 108 MB/s
BS: 8704 IOPS: 12956 BW: 108 MB/s
BS: 9216 IOPS: 12236 BW: 108 MB/s
BS: 9728 IOPS: 11592 BW: 108 MB/s
```

.8xlarge (one physical disk):
```
BS:  512 IOPS: 281474 BW: 137 MB/s
BS: 1024 IOPS: 280130 BW: 274 MB/s
BS: 1536 IOPS:  37882 BW:  55 MB/s
BS: 2048 IOPS: 277629 BW: 542 MB/s
BS: 2560 IOPS:  18968 BW:  46 MB/s
BS: 3072 IOPS:  18890 BW:  55 MB/s
BS: 3584 IOPS:  12615 BW:  43 MB/s
BS: 4096 IOPS: 277589 BW: 1084 MB/s
BS: 4608 IOPS: 254435 BW: 1118 MB/s
BS: 5120 IOPS: 253709 BW: 1239 MB/s
BS: 5632 IOPS: 250531 BW: 1346 MB/s
BS: 6144 IOPS: 254420 BW: 1491 MB/s
BS: 6656 IOPS: 241735 BW: 1534 MB/s
BS: 7168 IOPS: 250904 BW: 1715 MB/s
BS: 7680 IOPS: 232149 BW: 1700 MB/s
BS: 8192 IOPS: 230314 BW: 1799 MB/s
```

So you see some odd behaviour below 4k but also the additional step function at 7168 above which you reach the full boosted throughput. I don't think it's reasonable for iotune to try to reverse engineer all those points and try to force them into a model.

> Then let's correct it, 

> Criticism of the model is welcome, and even more if it comes with corrections.

Well I think the question comes a bit down to how much change/risk we would want. The advantage of the changes as in this PR are:

 - Improves perf quite a lot without fundamentally changing the model
 - As long as the IOs stay within min_iop_size/max_iop_size the model stays pessimistic such that scheduling keeps working even if you lose out some perf at some inflection point in that range

In terms of correcting the model I think it's just hard to generalize a model across all disks. The current model is probably good for that as it's fairly pessimistic in that sense.

Now on AWS for example (both dedicated disk and EBS) it seems as if everything behaves like a "max" model, i.e.: 

```math
cost = max(\frac{1}{\text{iops\_max}}, \frac{\text{size}}{\text{throughput\_max}})
```

So you get best performance (with good latency and scheduling) with a simple change like this:


```diff
diff --git a/src/core/io_queue.cc b/src/core/io_queue.cc
index 5f5ac7cc..c72482f1 100644
--- a/src/core/io_queue.cc
+++ b/src/core/io_queue.cc
@@ -833,7 +833,7 @@ double internal::request_tokens(io_direction_and_length dnl, const io_queue::con

     const auto& m = mult[dnl.rw_idx()];

-    return double(m.weight) / cfg.req_count_rate + double(m.size) * (dnl.length() >> io_queue::block_size_shift) / cfg.blocks_count_rate;
+    return std::max(double(m.weight) / cfg.req_count_rate,  double(m.size) * (dnl.length() >> io_queue::block_size_shift) / cfg.blocks_count_rate);
 }

 fair_queue_entry::capacity_t io_queue::request_capacity(io_direction_and_length dnl) const noexcept {
```

This seems to model the disks/controller fairly acurately and hence you get even better perf than with the improvements to the current model from this PR. For example, you get full perf and scheduling in the 128K "standard EBS" case and the 16k m8gd case. Note even this can't account for some of the weird sub 4k behaviour that is shown above but I don't think that's worth chasing.

But it's not clear whether the "max" model is a better default model for GCP, Azure or some random SSD out there. Possibly it is but I am not sure. Maybe the answer is just providing either with some io-properties flag.

avikivity commented:
> > rather than changing random parameters.
> 
> I mean the changes aren't entirely random. It's more like making the most out of the current model and measurements we already have.

If there's a reason why one number is better than the other, then I'll accept the one-liner patch with the two-page explanation.


> 
> > If there is indeed a behavioral change between 512 and 4096 bytes, we should extend the model to be a polygon rather than a straight line and measure the inflection points.
> 
> The true model model is likely some step function with many steps. Travis recently did some sub 4k testing and there seems to be quite varying behaviour. See for example on m8gd.large and m8gd.8xlarge (Travis has some more data but currently OOOish):

Is this read or write? I assume read, which shouldn't be sensitive to block size.

> 
> .large (partial disk):
> 
> ```
> BS: 512 IOPS: 16770 BW: 8 MB/s
> BS: 1024 IOPS: 16770 BW: 16 MB/s
> BS: 1536 IOPS: 16770 BW: 25 MB/s
> BS: 2048 IOPS: 16770 BW: 33 MB/s
> BS: 2560 IOPS: 16770 BW: 41 MB/s
> BS: 3072 IOPS: 16770 BW: 49 MB/s
> BS: 3584 IOPS: 12613 BW: 43 MB/s

Interesting. Could have been a noisy neighbor.

The numbers are otherwise too consistent, leading me to believe this is some QoS thing on a partial disk.

> BS: 4096 IOPS: 16770 BW: 66 MB/s
> BS: 4608 IOPS: 16770 BW: 74 MB/s
> BS: 5120 IOPS: 16770 BW: 82 MB/s
> BS: 5632 IOPS: 16770 BW: 90 MB/s
> BS: 6144 IOPS: 16770 BW: 98 MB/s
> BS: 6656 IOPS: 16770 BW: 106 MB/s
> BS: 7168 IOPS: 16335 BW: 112 MB/s
> BS: 7680 IOPS: 14684 BW: 108 MB/s
> BS: 8192 IOPS: 13765 BW: 108 MB/s
> BS: 8704 IOPS: 12956 BW: 108 MB/s
> BS: 9216 IOPS: 12236 BW: 108 MB/s
> BS: 9728 IOPS: 11592 BW: 108 MB/s
> ```
> 
> .8xlarge (one physical disk):
> 
> ```
> BS:  512 IOPS: 281474 BW: 137 MB/s
> BS: 1024 IOPS: 280130 BW: 274 MB/s

These are real numbers reported by real disks.

> BS: 1536 IOPS:  37882 BW:  55 MB/s
> BS: 2048 IOPS: 277629 BW: 542 MB/s
> BS: 2560 IOPS:  18968 BW:  46 MB/s
> BS: 3072 IOPS:  18890 BW:  55 MB/s
> BS: 3584 IOPS:  12615 BW:  43 MB/s

These sub-4k numbers are interesting and likely are related to crossing the block boundary. Are the 2k tests aligned on 2k boundary?

What is the logical/physical disk block sizes?

May be worthwhile reporting to AWS if we can't blame the kernel here.


> BS: 4096 IOPS: 277589 BW: 1084 MB/s
> BS: 4608 IOPS: 254435 BW: 1118 MB/s
> BS: 5120 IOPS: 253709 BW: 1239 MB/s
> BS: 5632 IOPS: 250531 BW: 1346 MB/s
> BS: 6144 IOPS: 254420 BW: 1491 MB/s
> BS: 6656 IOPS: 241735 BW: 1534 MB/s
> BS: 7168 IOPS: 250904 BW: 1715 MB/s
> BS: 7680 IOPS: 232149 BW: 1700 MB/s
> BS: 8192 IOPS: 230314 BW: 1799 MB/s
> ```
> 
> So you see some odd behaviour below 4k but also the additional step function at 7168 above which you reach the full boosted throughput. I don't think it's reasonable for iotune to try to reverse engineer all those points and try to force them into a model.

This is beyond a disk model and indicates a bug. There's no reason for 7680 to behave well while 3584 behaves badly (and no reason for 3584 to behave badly in the first place).


> 
> > Then let's correct it,
> 
> > Criticism of the model is welcome, and even more if it comes with corrections.
> 
> Well I think the question comes a bit down to how much change/risk we would want. The advantage of the changes as in this PR are:
> 
> * Improves perf quite a lot without fundamentally changing the model
> * As long as the IOs stay within min_iop_size/max_iop_size the model stays pessimistic such that scheduling keeps working even if you lose out some perf at some inflection point in that range

I'll have to go back and re-read since all the history drove the original patch out of my mind. But if we change to 4k then at least we need to change iotune to 4k, and we need to consider aligning iotune with it, and to check the disk block size.

> 
> In terms of correcting the model I think it's just hard to generalize a model across all disks. The current model is probably good for that as it's fairly pessimistic in that sense.
> 
> Now on AWS for example (both dedicated disk and EBS) it seems as if everything behaves like a "max" model, i.e.:
> 
> 𝑐
> 𝑜
> 𝑠
> 𝑡
> =
> 𝑚
> 𝑎
> 𝑥
> (
> 1
> iops_max
> ,
> size
> throughput_max
> )

EBS is indeed documented as max and since it's completely artificial (e.g. throttled) I trust it.

Local NVMe is additive. If you push maximum read IOPS you have to zero write bandwidth if you want good latency. If you push half max read IOPS you can push only half max write bandwidth if you want good performance.

> So you get best performance (with good latency and scheduling) with a simple change like this:
> 
> ```diff
> diff --git a/src/core/io_queue.cc b/src/core/io_queue.cc
> index 5f5ac7cc..c72482f1 100644
> --- a/src/core/io_queue.cc
> +++ b/src/core/io_queue.cc
> @@ -833,7 +833,7 @@ double internal::request_tokens(io_direction_and_length dnl, const io_queue::con
> 
>      const auto& m = mult[dnl.rw_idx()];
> 
> -    return double(m.weight) / cfg.req_count_rate + double(m.size) * (dnl.length() >> io_queue::block_size_shift) / cfg.blocks_count_rate;
> +    return std::max(double(m.weight) / cfg.req_count_rate,  double(m.size) * (dnl.length() >> io_queue::block_size_shift) / cfg.blocks_count_rate);
>  }
> 
>  fair_queue_entry::capacity_t io_queue::request_capacity(io_direction_and_length dnl) const noexcept {
> ```
> 
> This seems to model the disks/controller fairly acurately and hence you get even better perf than with the improvements to the current model from this PR. For example, you get full perf and scheduling in the 128K "standard EBS" case and the 16k m8gd case. Note even this can't account for some of the weird sub 4k behaviour that is shown above but I don't think that's worth chasing.
> 
> But it's not clear whether the "max" model is a better default model for GCP, Azure or some random SSD out there. Possibly it is but I am not sure. Maybe the answer is just providing either with some io-properties flag.

io-properties flag + code in iotune to discover it.

StephanDollberg commented:
> I'll have to go back and re-read since all the history drove the original patch out of my mind. But if we change to 4k then at least we need to change iotune to 4k, and we need to consider aligning iotune with it, and to check the disk block size.

In the commits as presented iotune will write out the min_iop_size as used during the tests to io-properties (TBD whether writes and reads should be treated separately). Right now this effectively always ends up being 512 because this is what the read IOPS test runs with. 512 is the min IO size (`minimum_io_size`) so it's really the lower limit. This makes this a safe default as for all IO sizes in between min_iop_size and max_iop_size the improved cost function just pessimistically approaches the existing one (you'll still get some improvements). 

However this doesn't limit users like RP who never use IOPs smaller than 4k and hence we wouldn't want to leave the additional performance go to waste and we would just run iotune with `--random-io-buffer-size=4096`. 

> If you push half max read IOPS you can push only half max write bandwidth if you want good performance. 

As an aside, I don't think it's a good idea to mix read and write workloads for this discussion as lots of the newer disks are actually somewhere in the spectrum between half and full duplex so when mixing reads and writes you do actually get higher total IOPS/throughput than when only doing reads or writes.

> Local NVMe is additive. If you push maximum read IOPS you have to zero write bandwidth if you want good latency. 

But this can't really be true if you think about it and is really why the current cost function doesn't work well. The moment you push some IOPS you push some throughput and vice versa.

In fact the SSDs seem to behave as if using two separate token buckets - one for IOPS and one for throughput. This is even more nuanced than the single max in the cost function that I showed above.

For example using the following io_tester config (without using io-properties) and testing on i3en.3xl (~250k rIOPS, ~2GB/s rTP) we get the following results:

```
- name: small_read
  shards: [0,1]
  type: randread
  shard_info:
    parallelism: 32
    reqsize: 4kB
    shares: 1000

- name: large_read
  shards: [2,3]
  type: randread
  shard_info:
    parallelism: 2
    reqsize: 128kB
    shares: 1000
```

abbreviated output:

```
---
- shard: 0
  small_read:
    throughput: 461184.531  # kB/s
    IOPS: 115296.633
    latencies:  # usec
      average: 276
      p0.5: 264
      p0.95: 411
      p0.99: 493
      p0.999: 650
      max: 1029
<cut>
- shard: 1
<cut - same as shard 0>
- shard: 2
  large_read:
    throughput: 628441.5  # kB/s
    IOPS: 4910.09912
    latencies:  # usec
      average: 406
      p0.5: 396
      p0.95: 546
      p0.99: 628
      p0.999: 741
      max: 947
<cut>
- shard: 3
<cut - same as shard 2>
...
```

So we are hitting the throughput limt of 2 * 628MB/s + 2 * 461 MB/s = ~2.2GB/s and at the same time we still do ~230k IOPS which is still more than 80% of max IOPS. All this at acceptable latency below the default io latency goal. Repeating the test with 512 IOP size for `small_read` and you run into the IOPS limit instead of the throughput limit. Repeating the test with `readwrite` shows a similar pattern. 

Now in practice workloads are a lot more mixed so you probably don't get the extreme ends. Also the min required IOP size for max throughput plays in. io depth likely as well.

When mixing write and read (i.e.: small_read and large_write) we get about 90% write throughput and 75% read IOPS. But again now we are talking about duplex behaviour which is another completely separate angle that heavily influences behaviour.

But anyway as touched on above the bigger problem in the io-queue is that the cost function itself is additive in terms of IOP and throughput cost. This costs each operation at a way too high cost and means that one can neither get full disk utilization in a 50/50 mixed workload described above or in the most basic one dimensional workload. Your total IOPS/throughput will stay short of what the disk can do and what is described in io-properties. This is what both the approach from the PR (by weighting the IOP and throughput cost) but also the "max" cost approach shown above (by costing a request based on the more expensive side of each) improve on.

StephanDollberg commented:
@xemul makes sense, btw I haven't pushed any changes yet as I see you guys are currently heavily refactoring io-queue related code.

Note I am not sure whether you have seen the discussion above about "max" cost function which works quite well in my testing on AWS and GCP and is a simpler change than what is proposed in the PR.

StephanDollberg committed:
io-queue: Refactor cost function

Pure non-functional refactor to split the long line into three.

(cherry picked from commit a0155baf6eb78c7c04436bf116174d487f962824)

StephanDollberg committed:
io-queue: Use max cost function by default

Switch the io-queue cost function over to use max(IOPS cost, TP cost)
away from sum(IOPS cost, TP cost).

In its form today the io-queue fails to reach the advertised
throughput/IOPS numbers from io-properties (see
[seastar#2771](https://github.com/scylladb/seastar/issues/2771) for
details).

The reason for this is easy to understand when looking at a disk like the
ones from the m7gd AWS series. They (approximately) achieve full throughput as
well as full IOPS at 4k IOP size.

This means that for cost modelling for a single IO operation it would be
enough to just look at one dimension, either throughput or IOPS but not
both. The current cost function which is always additive in IOP and throughput
cost however costs operation way too high. This leads to the aforementioned
problem where the io-queue underperforms. Taking 4k IOPS on
m7gd as an example we only get around 60% of the disk performance.

From various testing the max cost function seems to model disks better
than the sum cost function and avoids the problem mentioned above. Note
this doesn't change the "outer" cost model of how the costs of
individual operations are summed together. That also isn't modelled
entirely right for more modern disks but is pessimistically fine and not
the focus of this patch.

The problem can be easily shown using the following io-tester config:

```
- name: highprio
  shards: [0]
  type: randwrite
  shard_info:
    parallelism: 32
    reqsize: 4kB|16kB|128kB
    shares: 1000
    think_time: 0

- name: lowprio
  shards: [0]
  type: randwrite
  shard_info:
    parallelism: 32
    reqsize: 4kB|16kB|128kB
    shares: 100
    think_time: 0
```

On m7gd.4xl (all numbers in MB/s):
 - NOIP: Not using io-properties
 - IP: Using io-properties
 - MCF: Using io-properties with the max cost function

|       | 4k        |           |       | 16k       |           |       | 128k      |           |       |
|       | high prio | low prio  | total | high prio | low prio  | total | high prio | low prio  | total |
| NOIP  | 303       | 287       | 591   | 299       | 292       | 591   | 295       | 295       | 590   |
| IP    | 262       | 28        | 290   | 432       | 44        | 476   | 528       | 53        | 581   |
| MCF   | 476       | 56        | 532   | 543       | 55        | 598   | 544       | 55        | 599   |

We see that at low IOP size using no io-properties we do get the full
throughput but no scheduling (bad), using io-properties we do get scheduling
but low throughput and with the max cost function we get both.

More data and extensive discussion can be found here:

 - https://github.com/scylladb/seastar/pull/2801#issuecomment-3070114487
 - https://github.com/scylladb/seastar/issues/2840#issuecomment-3070100437

These tests also include latency tests in priority scenarios and
different kind of disks like EBS. The linked PR also includes a
discussion about a different approach using a weighted cost function.

Note that in general latency is often a function of throughput as the
io-queue arbitrarily keeping things in the queue induces latency. This
is especially true for real world usecases like RP where write
amplification is generally fairly high but at the same time can be
absorbed at the cost of latency.

(cherry picked from commit 9bf2094c446c85e2ff9890c31b0e42c3e6567271)

avikivity commented:
> @avikivity @xemul is there any ongoing interest in this change or should we assume that it will not be accepted at this point? We have applied the "MCF" approach described above to our fork with good results: it's configurable so either the existing or new MCF behavior is selectable.
> 
> We prefer to upstream as much as possible, usually "upstream first", but this win was too big to ignore. So just trying to gauge the mood here to see if this is a dead end here.

Don't interpret no response as lack of interest, I'm just lossy. Do ping me if I drop.

avikivity commented:
I'd like to check this with ScyllaDB. Please rebase it and make it work with old io_properties (assuming default iotune parameters). I'll try to get a test run.

xemul commented on include/seastar/core/io_queue.hh:
```diff
@@ -194,6 +194,7 @@ public:
         double flow_ratio_backpressure_threshold = 1.1;
         std::chrono::milliseconds stall_threshold = std::chrono::milliseconds(100);
         std::chrono::microseconds tau = std::chrono::milliseconds(5);
+        bool max_cost_function = true;
```
The default should be `false` to be backward compatible

xemul replied:
I tend to think that this approach needs improvement, or different justification. Here's my consideration, first some numerical example with odd numbers, then math description.


Let's take an imaginary disk with 1-kb/s bandwidth and 1-iops.
Let's say that request A is 2kb in size and request B is 0.5kb (512 bytes) in size.
Tokens rate, according to token-bucket implementation, is 1 per second (and 1 here is not because of disk bandwidth or IOPS, but just 1).
A's cost is `max(2/1, 1/1) = 2` with new model. B's cost is `max(0.5/1, 1/1) = 1`. 

Now let's saturate the queue with only A requests. 2 tokens are accumulated over 2 seconds, i.e. we'll see 1kb/s bandwidth and 0.5 iops. It's fine, it "fits" both -- the limitation and intention to get either maximum bandwidth or iops (here we get max bandwidth).

Next let's saturate the queue with only B requests. It will let one request in 1 second, we'll have 0.5kb/s bandwidth and 1iops, again -- everything is fine, the limitation is respected and the intention is achieved (max iops here).

Finally, let's put interleaved A and B into the queue, like this -- ABABABABA.... The AB pair cost is 3 tokens, which means that the bucket will let one AB pair in 3 seconds and we'll see the bandwidth of `2.5/3 ~= 0.8 kb/s` and iops of `2/3 ~= 0.3 iops`. This does fit the limitation, but the intent is broken -- neither maximum bandwidth nor iops is achieved, despite we expected it to happen.


I do agree that existing cost model will have its problems either. In particular A cost is 3 and B cost is 1.5 and neither 1kb/s nor 1iops is achievable for AAAAA, BBBB and ABABAB queues. My point here is different.

Defining cost to be `max(size/bw_max; 1/iops_max)` doesn't mean we're either reaching the maximum bandwidth or maximum iops. I only modeled equally mixed A and B queue, but what would happen if there's non-equal flow of A and B? Like AABAABAAB. Or even non-uniform like AABAAABBABAAB... Until this cost model is described with formulas, we are in risk to face unexpected side effects. When the token-bucket throttler was introduced we also assumed that it would be OK to replenish tokens not only by timer, but also tune this replenishment by completing requests. It then turned into disaster (#1766) and we had to roll it back and stick to analytically verified model.


As far as the formula is concerned. Current formula is

$$
    \frac {bw} B + \frac {iops} I \le 1.0
$$

Given linear property of any derivative wrt sum (that's important):

$$
    \frac d {dt} \( A + B \) = \frac {dA} {dt} + \frac {dB} {dt}
$$

and

$$
    bw = \frac d {dt} \( \sum bytes \) ; iops = \frac d {dt} \( \sum ops \)
$$

 we can turn the initial formula into

$$
    \frac d {dt} \( \frac {\sum bytes} B + \frac {\sum ops} I \) \le 1.0
$$

and it's valid to apply token bucket here with cost being $bytes + ops$

However, if we want to redefine cost as $max(bytes, ops)$ it will _not_ mean that the initial formula will be turned into

$$
    max \( \frac {bw} B ; \frac {iops} I \) \le 1.0
$$

simply because derivative is not linear wrt max():

$$
    \frac d {dt} max \( A ; B \) \ne max \( \frac {dA} {dt} ; \frac {dB} {dt} \)
$$

The initial formula will be something else.

xemul commented:
@StephanDollberg , the PR comment part about IOPS_IOP_SIZE and TP_IOP_SIZE is no longer relevant, isn't it?

avikivity commented:
@StephanDollberg please file a new pull request as my mistake [above](https://github.com/scylladb/seastar/pull/2801#issuecomment-3455859549) caused it to be marked closed.
